x/group/keeper/msg_server.go

```diff
@@ -756,6 +756,109 @@ func (k Keeper) Exec(goCtx context.Context, req *group.MsgExec) (*group.MsgExecR
 	return res, nil
 }
 
+// LeaveGroup implements the MsgServer/LeaveGroup method.
+func (k Keeper) LeaveGroup(goCtx context.Context, req *group.MsgLeaveGroup) (*group.MsgLeaveGroupResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	_, err := sdk.AccAddressFromBech32(req.MemberAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	groupInfo, err := k.getGroupInfo(ctx, req.GroupId)
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "group")
+	}
+
+	groupWeight, err := math.NewNonNegativeDecFromString(groupInfo.TotalWeight)
+	if err != nil {
+		return nil, err
+	}
+
+	gm, err := k.getGroupMember(ctx, &group.GroupMember{
+		GroupId: req.GroupId,
+		Member:  &group.Member{Address: req.MemberAddress},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	memberWeight, err := math.NewNonNegativeDecFromString(gm.Member.Weight)
+	if err != nil {
+		return nil, err
+	}
+
+	updatedWeight, err := math.SubNonNegative(groupWeight, memberWeight)
+	if err != nil {
+		return nil, err
+	}
+
+	iter, err := k.getGroupPoliciesByGroup(ctx, req.GroupId, nil)
+	if err != nil {
+		return nil, err
+	}
+	defer iter.Close()
+
+	var groupPolicy group.GroupPolicyInfo
+	for {
+		if _, err := iter.LoadNext(&groupPolicy); err != nil {
+			if errors.ErrORMIteratorDone.Is(err) {
+				break
+			}
+
+			return nil, err
+		}
+
+		policy := groupPolicy.GetDecisionPolicy()
+		tdp, ok := policy.(*group.ThresholdDecisionPolicy)
+		// TODO: handle percentage decision policy
+		if !ok {
+			return nil, sdkerrors.ErrInvalidRequest.Wrapf("expected %T, got %T", (*group.ThresholdDecisionPolicy)(nil), policy)
```

Comment:
shouldn't we just continue if it's not a ThresholdDecisionPolicy?

Reply:
Note: this makes it hard for external users to create their own policies that need weight-updating on members leaving. but maybe for now it's okay.